### PR TITLE
FORTA30: Victim & Loss Identifier bot

### PR DESCRIPTION
**Pull request by RCantu92**
# Victim & Loss Identifier bot

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] New Bot
- [ ] Fix a bot error
- [ ] Update a bot logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe):

## New Bot PR readiness

- [ ] `npm install` command works (i.e. package-lock is updated properly)
- [ ] The `name` and `description` fields of `package.json` describe the bot properly
- [ ] All the libraries listed in `package.json` are being used
- [ ] All the development libraries are in `devDependencies` section of `package.json`
- [ ] The bot README is updated
- [ ] The bot README sufficiently describes what the bot does
- [ ] All the alerts that the bot emits are listed in the README
- [ ] The txn/block examples listed in the README generate the expected findings
- [ ] Code is easy to follow and/or properly commented
- [ ] Bot contains tests
- [ ] All tests pass when running `npm test`
- [ ] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [ ] Tests cover all the paths (i.e. negative & positive test cases)
- [ ] When the tests complete, no warning/errors are shown
- [ ] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [ ] The networks specified in the README and the networks are listed in the `package.json` are the same
- [ ] The code does what the README describes
- [ ] Network calls are properly being cached (if duplicated calls can occur)
- [ ] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [ ] The findings fields are filled out correctly
- [ ] Findings `protocol` is specified (if the bot is for a specific one)
- [ ] `metadata` of the findings contains all the information needed.
- [ ] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [ ] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s
- [ ] The bot follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [ ] Bot can recover from failed error calls when they can fail (e.g. errors are being caught on calls that can fail)
- [ ] `repository` is added in the `package.json`

## Comments (optional)

If there is some useful information or important note for reviewing/understanding this bot please describe it here.

### CODE REVIEW

**Review by Vxatz — APPROVED** [review]

File: victim-loss-identifier/src/utils/utils.ts [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+  const fpData: FpTransaction[] = [];
+  const fpVictims: string[] = [];
+  const fpScammer = scammersCurrentlyMonitored[scammerAddress];
+
+  for (const victimAddress of Object.keys(fpScammer.victims!)) {
+    const victimInfo = fpScammer.victims![victimAddress];
+    const scammedBy = victimInfo.scammedBy;
+
+    const hasBeenAlerted = scammedBy[scammerAddress].hasBeenAlerted;
+    if (!hasBeenAlerted) continue;
```
Comment (RCantu92) [inline victim-loss-identifier/src/utils/utils.ts:65]:
```suggestion
    const hasBeenAlerted = scammedBy[scammerAddress].hasBeenAlerted;
    // Looking for victim's scammers that have had alerts
    // emitted to create an FP alert to "undo" the first one.
    // If they have no existing alert to "undo", skip
    if (!hasBeenAlerted) continue;
```
Suggestion to add a comment to explain this _peculiar_ `if` check.

File: victim-loss-identifier/src/utils/utils.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+            victims[victimAddress].totalUsdValueAcrossAllErc721Tokens! -= erc721TokenInfo.tokenTotalUsdValue!;
+          }
+          // Format ERC721 data as "tokenId,tokenAddress"
+          tokenIds.forEach((tokenId) => {
+            extractedTransaction.nfts.push(`${tokenId},${tokenAddress}`);
+          });
+        });
+      }
+
+      if (transaction.erc1155) {
```
Comment (RCantu92) [inline victim-loss-identifier/src/utils/utils.ts:92]: Is this check to be more "future proof"? As in to already have in mind that there will be cases were a victim could be scammed of their ERC1155s? _Or_, is the bot supposed to handle the case of ERC1155 already?

If it is the latter, then the functions `addVictimInfoToVictims` and `increaseStolenUsdAmounts` in `fraud.nft.order.processing.ts` need to be updated to properly handle ERC1155 exploit instances.

Comment (RCantu92) [reply victim-loss-identifier/src/utils/utils.ts:92]: If it is the former, perhaps adding the logic to ERC20 transactions should be added.

WDYT?

**Review by RCantu92 — COMMENTED** [review]
Mostly minor comments, except for the `transaction.erc1155` check comment.

File: victim-loss-identifier/src/utils/utils.ts (hunk 2 above)
Comment (Vxatz) [reply victim-loss-identifier/src/utils/utils.ts:92]: @RCantu92, it's the former, but ERC20s are not expected to generate a Label at all, I'd say, what do you think?

Comment (Vxatz) [reply victim-loss-identifier/src/utils/utils.ts:92]: E.g. in the Ice Phishing we're only labelling NFTs and I think that make sense due to ERC20s being totally fungible

File: victim-loss-identifier/src/fetcher.ts [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+              return current;
+            }
+            return latest;
+          },
+          {
+            floorPrice: 0,
+            priceCurrency: "",
+            collectionUrl: "",
+            retrievedAt: "2000-01-01T00:00:00.000Z",
+          }
```
[the remark this answers is not in the document]
Comment (Vxatz) [reply victim-loss-identifier/src/fetcher.ts:268]: You're right. I thought of it as a "good practice" but yeah it doesn't make sense tbh.

Also added a "not null" check:

```
 if (previousEntry && previousEntry.priceCurrency !== "ETH") {
          return 0;
        }

        return previousEntry ? previousEntry.floorPrice : 0;
```

**Pull request merged**